nixos/modules/services/hardware/nvidia-container-toolkit-cdi-generator/cdi-generate.nix

```diff
@@ -0,0 +1,37 @@
+{ config, lib, pkgs }: let
+  mountOptions = { options = ["ro" "nosuid" "nodev" "bind"]; };
+  mounts = [
+    { hostPath = "${lib.getBin config.hardware.nvidia.package}/bin/nvidia-cuda-mps-control";
+      containerPath = "/usr/bin/nvidia-cuda-mps-control"; }
+    { hostPath = "${lib.getBin config.hardware.nvidia.package}/bin/nvidia-cuda-mps-server";
+      containerPath = "/usr/bin/nvidia-cuda-mps-server"; }
+    { hostPath = "${lib.getBin config.hardware.nvidia.package}/bin/nvidia-debugdump";
+      containerPath = "/usr/bin/nvidia-debugdump"; }
+    { hostPath = "${lib.getBin config.hardware.nvidia.package}/bin/nvidia-powerd";
+      containerPath = "/usr/bin/nvidia-powerd"; }
+    { hostPath = "${lib.getBin config.hardware.nvidia.package}/bin/nvidia-smi";
+      containerPath = "/usr/bin/nvidia-smi"; }
+    { hostPath = "${pkgs.nvidia-container-toolkit}/bin/nvidia-ctk";
+      containerPath = "/usr/bin/nvidia-ctk"; }
+    { hostPath = "${pkgs.glibc}/lib";
+      containerPath = "${pkgs.glibc}/lib"; }
```

Comment:
Aha! So, this has nothing to do with interfaces so I won't call this a blocker.

However, I presume the reason you're mounting this is because the other mounted programs (`nvidia-smi`) depend on this libc? In that case, this only "works" _accidentally_ (generally, these programs might have been overridden to link against a different glibc, etc). We also shouldn't be wasting time tracking this kind of transitive dependencies manually. Note that the list of transitive runtime dependencies cannot be generated at eval time, but requires a build (e.g. `pkgs.closureInfo`).


As I said, I'd consider this an implementation detail and the proper solution can be implemented in a subsequent PR

Reply:
I think the only thing that `cdi.dynamic.nvidia.enable` should be adding to the ldconfig cache is the nvidia driver libraries. Libc&c should only be mounted _optionally_. Generally speaking, mounting host's libc into the container is very much the wrong thing to do (containers may come with their own libc).

AFAIU this only gets mounted when the nvidia generator is enabled. Is that correct?

Reply:
> AFAIU this only gets mounted when the nvidia generator is enabled. Is that correct?

Yes, exactly. This is the script that generates the `/run/cdi/nvidia-container-toolkit.json`, that is only present when the nvidia generator is enabled.